### DISCUSSION TITLE
[cloud_firestore] DocumentSnapshot promotion or null safe `data`

### DISCUSSION THREAD

**Discussion by stargazing-dino** [firebase/flutterfire#6060] [Q&A] [answered]
Hi ! Currently, a `_WithConverterDocumentSnapshot.data()` has return type of a `T?` but wouldn't it be better for the nullability to be defined by the user? So just make it `T` and let the say user if it's ever going to be a type `T?`

So like turning this

```dart
@override
T? data() {
  if (!_originalDocumentSnapshot.exists) return null;

  return _fromFirestore(_originalDocumentSnapshot, null);
}

@override
bool get exists => _originalDocumentSnapshot.exists;
```

into something like

```dart
@override
T data() {
  if (!_originalDocumentSnapshot.exists) {
    // Check if it's nullable
    if (null is T) {
      return null;
    } else {
      throw 'Non-nullable type given but no snapshot does not exist.';
    }
  }

  return _fromFirestore(_originalDocumentSnapshot, null);
}

@override
bool get exists {
  if (null is T) {
    return _originalDocumentSnapshot.exists;
  } else {
    return true;
  }
}
```

I ran into this because I was hoping to somehow promote `DocumentSnapshot` to `QueryDocumentSnapshot` to get the data without force unwrapping but I couldn't find a proper way to.

If this is not possible could there be a separate method to get `T?` and `T` similar to `context.maybeOf` and `context.of`

cc @rrousselGit  (Sorry for pinging!)

**Comment by rrousselGit** [comment, marked as answer]
You should be able to implement that with an extension:

```dart
extension<T> on DocumentSnapshot<T> {
  T get requiredData {
     final value = data();
     if (!value.exists) throw Error('not found');
     return value as T;
  }
}
```

**Reply by stargazing-dino** [reply]
No auto imports on extensions have kept me from liking them but that sounds like a good solution for me now

Thanks Remi :)
